internal/server/polling/server.go

```diff
@@ -0,0 +1,107 @@
+package polling
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/weaveworks/tf-controller/internal/git/provider"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const DefaultPollingInterval = time.Second * 30
+
+type Server struct {
+	log             logr.Logger
+	clusterClient   client.Client
+	configMapRef    client.ObjectKey
+	pollingInterval time.Duration
+}
+
+func New(options ...Option) (*Server, error) {
+	server := &Server{log: logr.Discard()}
+
+	for _, opt := range options {
+		if err := opt(server); err != nil {
+			return nil, err
+		}
+	}
+
+	return server, nil
+}
+
+func (s *Server) Start(ctx context.Context) error {
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		// Read the config in each iteration. The idea behind this decision is to
+		// allow the user to change the list of resources without the need of
+		// restart of the pod.
+		// It can be a bit smarter like using a time.Ticker and refresh config
+		// periodically.
+		config, err := s.readConfig(ctx)
+		if err != nil {
+			return err
+		}
+
+		secret, err := s.getSecret(ctx, client.ObjectKey{
+			Namespace: config.SecretNamespace,
+			Name:      config.SecretName,
+		})
+		if err != nil {
+			s.log.Error(err, "failed to get secret")
+		}
+
+		for _, resource := range config.Resources {
+			if err := s.poll(ctx, resource, secret); err != nil {
+				s.log.Error(err, "failed to check pull request")
+			}
+		}
+
+		time.Sleep(s.pollingInterval)
+	}
+}
+
+func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret *corev1.Secret) error {
+	if secret == nil {
+		return fmt.Errorf("secret is not defined")
+	}
+
+	tf, err := s.getTerraform(ctx, resource)
+	if err != nil {
+		return fmt.Errorf("failed to get Terraform object: %w", err)
+	}
+
+	source, err := s.getSource(ctx, tf)
+	if err != nil {
+		return fmt.Errorf("failed to get Source object: %w", err)
+	}
+
+	gitProvider, repo, err := provider.FromURL(
+		source.Spec.URL,
+		provider.WithLogger(s.log),
+		provider.WithToken("api-token", string(secret.Data["token"])),
```

Comment:
yes and there is a [constant for that](https://github.com/weaveworks/tf-controller/blob/main/internal/git/provider/options.go#L7) 🤦‍♀️ 

Sadly, long running PRs has this very bad side effect, it can be a mess at the end.

Reply:
:nods: